inc/elasticsearch_packages/namespace.php

```diff
@@ -0,0 +1,432 @@
+		$recheck_statuses = [
+			'COPYING',
+			'ASSOCIATING',
+			'VALIDATING',
+		];
+		if ( ! in_array( $status, $recheck_statuses, true ) ) {
+			$error_message = sprintf( 'Package %s has encountered an error. See the status for more details.', $package_id );
+			throw new Exception( $error_message );
+		}
+	} catch ( Exception $e ) {
+		// Unschedule this hook.
+		$next = wp_next_scheduled( 'altis.search.check_package_status', $check_status_hook_args );
+		if ( $next ) {
+			wp_unschedule_event( $next, 'altis.search.check_package_status', $check_status_hook_args );
+		}
+
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		trigger_error( $e->getMessage(), E_USER_WARNING );
+	}
+}
+
+/**
+ * Loop over packages to remove and delete them. Called on
+ * the altis.search.updated_all_index_settings hook.
+ *
+ * @return void
+ */
+function on_update_index_settings() {
+	$packages_to_remove = get_packages_to_remove();
+
+	foreach ( $packages_to_remove as $package_id ) {
+		delete_package( $package_id );
+	}
+
+	// Empty the array of packages to be removed.
+	update_site_option( 'altis_search_packages_to_remove', [] );
+}
+
+/**
+ * Hook triggered when deleting a package directly.
+ *
+ * @param string $package_id The old package ID that has been deleted.
+ */
+function on_deleted_package( string $package_id ) {
+	delete_package( $package_id );
+}
+
+/**
+ * Delete a package off Elasticsearch Service.
+ *
+ * @param string $package_id The package ID to delete.
+ * @param int $max_retries The maximum number of times to attempt deletion.
+ * @return void
+ */
+function delete_package( string $package_id, int $max_retries = 5 ) {
+
+	$client = get_es_client();
+
+	$real_package_id = str_replace( 'analyzers/', '', $package_id );
+
+	// Check if we've hit the retry limit.
+	if ( $max_retries === 0 ) {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		trigger_error( sprintf( 'Error deleting package %s on Elasticsearch Service', $real_package_id ), E_USER_WARNING );
+		return;
+	}
+
+	try {
+		// First check the package is no longer associated with the domain.
+		$packages = $client->listDomainsForPackage( [
+			'PackageID' => $real_package_id,
+		] );
+
+		if ( ! empty( $packages['DomainPackageDetailsList'] ) ) {
+			// Package is currently being associated or is active.
+			$package = $packages['DomainPackageDetailsList'][0];
+			$status = $package['DomainPackageStatus'];
+
+			// Dissociate the package if it's currently active or dissociation failed.
+			if ( $status === 'DISSOCIATION_FAILED' || $status === 'ACTIVE' ) {
+				$client->dissociatePackage( [
+					'DomainName' => get_elasticsearch_domain(), // required.
+					'PackageID' => $real_package_id, // required.
+				] );
+			}
+
+			// Reschedule deletion attempt.
+			wp_schedule_single_event( time() + ( 5 * MINUTE_IN_SECONDS ), 'altis.search.delete_package', [
+				$package_id,
+				$max_retries - 1,
+			] );
+
+			return;
```

Comment:
Should we return a code to indicate to the consumer whether deletion occurred vs was scheduled vs failed?

Reply:
Not sure where that would be used / useful right now. This is a background task.

Reply:
That's true, but the function itself (`delete_package()`) seems like a generic API that anyone could use internally?

Reply:
Yeah that's true. Makes me wonder if there's a nicer way async functions like this could be packaged. Feels quite separate from the action hooks required to make it work.